python-frontend/src/test/java/org/sonar/python/types/InferredTypesTest.java

```diff
@@ -282,6 +306,62 @@ void test_optional_type_annotations() {
       .containsExactlyInAnyOrder("typing.Optional");
   }
 
+  @Test
+  void test_union_type_annotation_with_builtins() {
+    DeclaredType union = new DeclaredType(new SymbolImpl("Union", "typing.Union"), List.of(new DeclaredType(BuiltinTypes.INT), new DeclaredType(BuiltinTypes.STR)));
+    assertThat(lastExpression(
+      "def f(x: int | str):",
+      "  x").type()).isEqualTo(union);
+  }
+
+  @Test
+  void test_union_type_annotation_inside_generics() {
+    DeclaredType union = new DeclaredType(new SymbolImpl("Union", "typing.Union"), List.of(new DeclaredType(BuiltinTypes.INT), new DeclaredType(BuiltinTypes.STR)));
+    DeclaredType listOfUnion = new DeclaredType(new SymbolImpl(BuiltinTypes.LIST, BuiltinTypes.LIST), List.of(union));
+    assertThat(lastExpression(
+      "def f(x: list[int | str]):",
+      "  x").type()).isEqualTo(listOfUnion);
+  }
+
+  @Test
+  void test_union_type_annotation_with_unknown_types() {
+    assertThat(lastExpression(
+      "def f(x: int | A):",
+      "  x").type()).isEqualTo(anyType());
+
+    assertThat(lastExpression(
+      "def f(x: B | int):",
+      "  x").type()).isEqualTo(anyType());
+  }
+
+  @Test
+  void test_union_type_annotation_with_any_type() {
```

Comment:
Maybe the title of this test is a little bit misleading, because we are not testing a uniton type here.